jax/_src/dtypes.py

```diff
@@ -343,7 +343,9 @@ def is_python_scalar(x):
 
 def dtype(x, *, canonicalize=False):
   """Return the dtype object for a value or type, optionally canonicalized based on X64 mode."""
-  if isinstance(x, type) and x in python_scalar_dtypes:
+  if x is None:
```

Comment:
I'm wondering if callers expecting this behavior should be calling an API from `jnp` instead. Should `dtypes` encode this particular numpy quirk?

Reply:
`dtypes` already encodes this particular quirk, because if `x = None` then it hits `np.result_type(x)`, which returns `float64`.

I made this change because I thought it would be better for `None` to explicitly return JAX's default float dtype, rather than implicitly returning numpy's default float dtype (currently the two defaults match, which is why this does not change the behavior of the function, but for the sake of #8178 returning the JAX default is useful)

I think it would be useful to make `x=None` result in a `ValueError` within `lax.dtype` and `lax.convert_element_type`, but since that would be a breaking change I think it would make sense to do in a separate PR.